### PR TITLE
add document for GenAI

### PR DESCRIPTION
**Pull request by LittleLittleCloud**
We are excited to review your PR.

So we can do the best job, please check:

- [x] There's a descriptive title that will make sense to other developers some time from now. 
- [x] There's associated issues. All PR's should have issue(s) associated - unless a trivial self-evident change such as fixing a typo. You can use the format `Fixes #nnnn` in your description to cause GitHub to automatically close the issue(s) when your PR is merged.
- [x] Your change description explains what the change does, why you chose your approach, and anything else that reviewers should know.
- [ ] You have included any necessary tests in the same PR.
- [ ] 
Add desgin document for #7169

### CODE REVIEW

**Comment by luisquintanilla** [comment]
@NiklasGustafsson @stephentoub @SteveSandersonMS can you please take a look and provide feedback.

This is the design document for the GenAI library which consolidates the models ported to TorchSharp (Phi, Llama, StableDiffusion) into a single library for easy consumption.

File: docs/gen-ai/CausalLMPipeline.md [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+        Tensor attentionMask, // attention mask [batch_size, sequence_length]
+        float temperature = 0.7f,
+        float topP = 0.9f,
+        int maxLen = 128,
+        int[][]? stopTokenSequence = null,
+        bool echo = false); // echo the input token ids in the output token ids
+}
+
+public CasualLMPipeline<TTokenizer, TCausalLM> : CausalLMPipeline
+    where TTokenizer : ITokenizer
```
Comment (stephentoub) [inline docs/gen-ai/CausalLMPipeline.md:24]: How does this relate to Microsoft.ML.Tokenizers? Can this just be Tokenizer?

File: docs/gen-ai/CausalLMPipeline.md [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+The causal language model pipeline is a utility class which wraps a tokenizer and a causal language model and provides a uniformed interface for various decoding method to generate text. The pipeline is designed to be easy to use and requires only a few lines of code to generate text.
+
+In Microsoft.ML.GenAI, we will provide a generic `CausalLMPipeline` class plus a typed `CausalLMPipeline` class which specifies the type parameters for the tokenizer and the causal language model. The typed `CausalLMPipeline` class make it easier to develop consuming method for semantic kernel. see [here](./Usage.md#consume-model-from-semantic-kernel) for more details
+# Contract
+```C#
+public abstract class CausalLMPipeline
+{
+    public virtual (
+        Tensor, // output token ids [batch_size, sequence_length]
+        Tensor // output logits [batch_size, sequence_length, vocab_size]
```
Comment (stephentoub) [inline docs/gen-ai/CausalLMPipeline.md:12]: Is this the TorchSharp tensor type? Can / should it become the System.Numerics.Tensors one?

File: docs/gen-ai/CausalLMPipeline.md [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+        int maxLen = 128,
+        int[][]? stopTokenSequence = null,
+        bool echo = false); // echo the input token ids in the output token ids
+}
+
+public CasualLMPipeline<TTokenizer, TCausalLM> : CausalLMPipeline
+    where TTokenizer : ITokenizer
+    where TCausalLM : nn.Module<CausalLanguageModelInput, CausalLanguageModelOutput>
+{
+    public CausalLMPipeline<LLama2Tokenizer, Phi3ForCasualLM> Create(LLama2Tokenizer tokenizer, Phi3ForCasualLM model);
```
Comment (stephentoub) [inline docs/gen-ai/CausalLMPipeline.md:27]: Presumably this is intended to instead be a static method on the base class?

File: docs/gen-ai/Package Structure.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+The GenAI project will be a collection of popular open source AI models. It will be organized in the following structure:
```
Comment (stephentoub) [inline docs/gen-ai/Package Structure.md:1]: Who would we direct to use this package vs the ONNX runtime GenAI package? e.g. in what situation would someone be better off using this one instead of that one? Models that it doesn't support that this one does?  Or is this more for experimentation / samples rather than us actually publishing a nuget package we'd encourage folks to adopt? (Which would be fine, just trying to understand the landscape.)

File: docs/gen-ai/Tokenizer.md [hunk 5]
```diff
@@ -0,0 +1,6 @@
+# What is a tokenizer?
+
+A tokenizer is a class that splits a string into tokens and encodes them into numerical(int) values.
+
+# The Tokenizer Contract
+We can simply use the tokenizer from `Microsoft.ML.Tokenizer` package
```
Comment (stephentoub) [inline docs/gen-ai/Tokenizer.md:6]: Great

File: docs/gen-ai/Usage.md [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+
+var prompt = "<|user|>Once upon a time<|end|><assistant>";
+var output = pipeline.Generate(
+    prompt: prompt,
+
+    maxLen: 100);
+```
+
+### Consume model from semantic kernel
+In most cases, developers would like to consume the model in a uniformed way. In this case, we can provide an extension method to semantic kernel which adds CausalLMPipeline as `ChatCompletionService`
```
Comment (stephentoub) [inline docs/gen-ai/Usage.md:22]: Great

File: docs/gen-ai/Usage.md [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+var tokenizer = LLama2Tokenizer.FromPretrained(pathToPhi3);
+var phi3CausalModel = Phi3ForCasualLM.FromPretrained(pathToPhi3);
+var pipeline = new CausalLMPipeline(tokenizer, phi3CausalModel);
+var agent = new Phi3MiniAgent(pipeline, name: "assistant");
+
+// AutoGen.Net allows you to run the agent as an OpenAI chat completion endpoint
+var host = Host.CreateDefaultBuilder()
+    .ConfigureWebHostDefaults(app =>
+    {
+        app.UseAgentAsOpenAIChatCompletionEndpoint(agent);
```
Comment (stephentoub) [inline docs/gen-ai/Usage.md:70]: Is this from autogen?

Comment (LittleLittleCloud) [reply docs/gen-ai/Usage.md:70]: Not for now, It's a WIP PR from AutoGen.Net

https://github.com/microsoft/autogen/pull/2633

File: docs/gen-ai/Package Structure.md (hunk 4 above)
Comment (luisquintanilla) [reply docs/gen-ai/Package Structure.md:1]: > Who would we direct to use this package vs the ONNX runtime GenAI package?

ORT GenAI package is what we'd recommend if all you need is inferencing. As far as I know though there are no plans at this time to enable fine-tuning in that library. While the scenario this package currently enables is inferencing just like ORT GenAI, having a version of these models in TorchSharp would enable fine-tuning scenarios as well. Currently in ML.NET, for [object detection](https://devblogs.microsoft.com/dotnet/object-detection-ml-dotnet-model-builder/#object-detection-in-model-builder) and [language scenarios](https://devblogs.microsoft.com/dotnet/introducing-the-ml-dotnet-text-classification-api-preview/), TorchSharp is what powers the fine-tuning of those models. As a next step for this package, we'd be looking to enable fine-tuning. 

> Models that it doesn't support that this one does

In terms of model coverage there is overlap. However, ORT Gen AI provides access to other language models that this library doesn't.

image

> experimentation / samples rather than us actually publishing a nuget package we'd encourage folks to adopt?

Initially this would be an experimental package as we get feedback and enable fine-tuning capabilities.

Comment (LittleLittleCloud) [reply docs/gen-ai/Package Structure.md:1]: They target for different model format. This GenAI packages loads the same weight as pytorch model. (.safetensor or .bin) while onnx runtime GenAI loads .onnx model format, which usually requires an extra converting step.

So if a model provides both onnx or .safetensor model weight format, then it's up to you to either use this GenAI package or the onnx runtime GenAI package. In most cases however, you need to convert .safetensor or .bin to .onnx by yourself or download 3p onnx model weight from enthusiastic community members before using ONNX runtime GenAI package.

File: docs/gen-ai/CausalLMPipeline.md (hunk 2 above)
Comment (LittleLittleCloud) [reply docs/gen-ai/CausalLMPipeline.md:12]: > Is this the TorchSharp tensor type

Yes

> should it become the System.Numerics.Tensors one?
If System.Numerics.Tensors can work with libtorch, then maybe yes.

File: docs/gen-ai/CausalLMPipeline.md (hunk 1 above)
Comment (LittleLittleCloud) [reply docs/gen-ai/CausalLMPipeline.md:24]: Can

File: docs/gen-ai/CausalLMPipeline.md (hunk 3 above)
Comment (LittleLittleCloud) [reply docs/gen-ai/CausalLMPipeline.md:27]: Hmmm on the second thought, maybe not. Because `CasualLMPipeline` and `CasualLMPipeline<TTokenizer, TCasualLM>` will live in `GenAI.Core` while `Phi3ForCasualLM` lives in `GenAI.Phi3`.

In that case, this `Create` method is no longer necessary exist, using a constructor to take `TTokenzier` and `TCausalLM` would be easier to understand

File: docs/gen-ai/CausalLMPipeline.md (hunk 2 above)
Comment (NiklasGustafsson) [reply docs/gen-ai/CausalLMPipeline.md:12]: We're some ways from having Tensor<T> integrated with TorchSharp, it is still experimental and subject to change.

Comment (NiklasGustafsson) [reply docs/gen-ai/CausalLMPipeline.md:12]: And it would be the TorchSharp shim implementing ITensor<T>, not Tensor<T> from SNT.

Comment (stephentoub) [reply docs/gen-ai/CausalLMPipeline.md:12]: Presumably it would be `ITensor<T>` such that a `Tensor<T>` could be passed in? The implementation could type test if it were a TorchSharp implementation and act accordingly, presumably.

Comment (NiklasGustafsson) [reply docs/gen-ai/CausalLMPipeline.md:12]: Yes, in the fullness of time.

**Review by luisquintanilla — APPROVED** [review]
LGTM
